src/thread.rs

```diff
@@ -0,0 +1,611 @@
+    }
+
+    /// Enable the thread for execution. The thread must be terminated.
+    fn enable_thread(&mut self, thread_id: ThreadId) {
+        assert!(self.has_terminated(thread_id));
+        self.threads[thread_id].state = ThreadState::Enabled;
+    }
+
+    /// Get a mutable borrow of the currently active thread.
+    fn active_thread_mut(&mut self) -> &mut Thread<'mir, 'tcx> {
+        &mut self.threads[self.active_thread]
+    }
+
+    /// Get a shared borrow of the currently active thread.
+    fn active_thread_ref(&self) -> &Thread<'mir, 'tcx> {
+        &self.threads[self.active_thread]
+    }
+
+    /// Mark the thread as detached, which means that no other thread will try
+    /// to join it and the thread is responsible for cleaning up.
+    fn detach_thread(&mut self, id: ThreadId) -> InterpResult<'tcx> {
+        if self.threads[id].join_status != ThreadJoinStatus::Joinable {
+            throw_ub_format!("trying to detach thread that was already detached or joined");
+        }
+        self.threads[id].join_status = ThreadJoinStatus::Detached;
+        Ok(())
+    }
+
+    /// Mark that the active thread tries to join the thread with `joined_thread_id`.
+    fn join_thread(&mut self, joined_thread_id: ThreadId) -> InterpResult<'tcx> {
+        if self.threads[joined_thread_id].join_status != ThreadJoinStatus::Joinable {
+            throw_ub_format!("trying to join a detached or already joined thread");
+        }
+        if joined_thread_id == self.active_thread {
+            throw_ub_format!("trying to join itself");
+        }
+        assert!(
+            self.threads
+                .iter()
+                .all(|thread| thread.state != ThreadState::BlockedOnJoin(joined_thread_id)),
+            "a joinable thread already has threads waiting for its termination"
+        );
+        // Mark the joined thread as being joined so that we detect if other
+        // threads try to join it.
+        self.threads[joined_thread_id].join_status = ThreadJoinStatus::Joined;
+        if self.threads[joined_thread_id].state != ThreadState::Terminated {
+            // The joined thread is still running, we need to wait for it.
+            self.active_thread_mut().state = ThreadState::BlockedOnJoin(joined_thread_id);
+            trace!(
+                "{:?} blocked on {:?} when trying to join",
+                self.active_thread,
+                joined_thread_id
+            );
+        }
+        Ok(())
+    }
+
+    /// Set the name of the active thread.
+    fn set_thread_name(&mut self, new_thread_name: Vec<u8>) {
+        self.active_thread_mut().thread_name = Some(new_thread_name);
+    }
+
+    /// Get the name of the active thread.
+    fn get_thread_name(&self) -> InterpResult<'tcx, &[u8]> {
+        if let Some(ref thread_name) = self.active_thread_ref().thread_name {
+            Ok(thread_name)
+        } else {
+            throw_ub_format!("thread {:?} has no name set", self.active_thread)
```

Comment:
I don't see any indication in the docs that getting the name of a thread before setting it is UB. Also, debug-printing the thread name here is not terribly useful as it is not set. ;)

So what about just returning `b"<unnamed>"` or so?

Reply:
> Also, debug-printing the thread name here is not terribly useful as it is not set. ;)

This is debug-printing the thread id, not thread name.